src/main/java/org/springframework/integration/kafka/inbound/KafkaInboundGateway.java

```diff
@@ -345,7 +345,7 @@ public void onMessage(ConsumerRecord<K, V> record, Acknowledgment acknowledgment
 
 		@Override
 		public <T, E extends Throwable> boolean open(RetryContext context, RetryCallback<T, E> callback) {
-			if (KafkaInboundGateway.this.recoveryCallback != null) {
+			if (KafkaInboundGateway.this.retryTemplate != null) {
```

Comment:
May we have some description why so critical change?
Looks like you just broke a logic to facilitate your requirements.

Thanks.

Reply:
1. if you look at occurrences of attributesHolder in the same file they are protected by a null check on retryTemplate not on recoveryCallback, resulting in an NPE if recoverycallback is null but retryTemplate is not. This seems weird to me.
2. It seems to me that this is a change in behaviour between version 3.1.2 and 3.1.3
3. If the recoveryCallback is mandatory, the code should fail early, right now it has an NPE in #onMessage

Reply:
I'm not sure what you mean about versions, but such a behavior was there from the beginning: https://github.com/spring-projects/spring-integration-kafka/commit/1de7b67013cca1e5f9cd6e54e49709c90553385b#diff-b49c842da794eb3622242c8acb03e1c1R385

you might be right, of course, but it would be grate to be sure that we are on the same page. 

Also I'm not sure why you issue PR against version `3.1.x`. How about a `master` or even that `3.0.x` we have introduced this problem?

Reply:
Does your test get the NPE without the source changes?

An NPE is definitely bad, but the logic is as intended.

We don't need to add the attributes to the retry context unless there is a recovery callback; it's the callback that needs the input message etc.

When there is no recovery callback, the retry template throws the exception after the retries are exhausted and we call `sendErrorMessageIfNecessary` which checks to see if there is an `errorChannel`.

Can you provide a stack trace for the NPE so we can see what's going on?

Reply:
Yes if you run the additional tests on otherwise unmodified v3.1.4.RELEASE they fail, they pass on v3.1.3.RELEASE.

Reply:
Stacktrace 
```
java.lang.NullPointerException
	at org.springframework.integration.kafka.inbound.KafkaMessageDrivenChannelAdapter$IntegrationRecordMessageListener.enhanceHeaders(KafkaMessageDrivenChannelAdapter.java:450)
	at org.springframework.integration.kafka.inbound.KafkaMessageDrivenChannelAdapter$IntegrationRecordMessageListener.onMessage(KafkaMessageDrivenChannelAdapter.java:433)
	at org.springframework.integration.kafka.inbound.KafkaMessageDrivenChannelAdapter$IntegrationRecordMessageListener.onMessage(KafkaMessageDrivenChannelAdapter.java:415)
	at org.springframework.kafka.listener.adapter.RetryingMessageListenerAdapter.lambda$onMessage$0(RetryingMessageListenerAdapter.java:120)
	at org.springframework.retry.support.RetryTemplate.doExecute(RetryTemplate.java:287)
	at org.springframework.retry.support.RetryTemplate.execute(RetryTemplate.java:211)
	at org.springframework.kafka.listener.adapter.RetryingMessageListenerAdapter.onMessage(RetryingMessageListenerAdapter.java:114)
	at org.springframework.kafka.listener.adapter.RetryingMessageListenerAdapter.onMessage(RetryingMessageListenerAdapter.java:40)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.doInvokeOnMessage(KafkaMessageListenerContainer.java:1275)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.invokeOnMessage(KafkaMessageListenerContainer.java:1258)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.doInvokeRecordListener(KafkaMessageListenerContainer.java:1219)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.doInvokeWithRecords(KafkaMessageListenerContainer.java:1200)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.invokeRecordListener(KafkaMessageListenerContainer.java:1120)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.invokeListener(KafkaMessageListenerContainer.java:935)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.pollAndInvoke(KafkaMessageListenerContainer.java:751)
	at org.springframework.kafka.listener.KafkaMessageListenerContainer$ListenerConsumer.run(KafkaMessageListenerContainer.java:700)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:264)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.base/java.lang.Thread.run(Thread.java:834)

```

Reply:
Only this change happens in between: https://github.com/spring-projects/spring-integration-kafka/commit/90eb5d1686c58c8f283a8c6f091e6b3b351ea17c.
See commit history: https://github.com/spring-projects/spring-integration-kafka/commits/v3.1.4.RELEASE

So, we may have an NPE, but I'm not sure how is that related to the code your are changing...

Reply:
> at org.springframework.integration.kafka.inbound.KafkaMessageDrivenChannelAdapter$IntegrationRecordMessageListener.enhanceHeaders(KafkaMessageDrivenChannelAdapter.java:450)

```
if (KafkaMessageDrivenChannelAdapter.this.retryTemplate != null) {
					AtomicInteger deliveryAttempt =
							new AtomicInteger(((RetryContext) attributesHolder.get()).getRetryCount() + 1);
					rawHeaders.put(IntegrationMessageHeaderAccessor.DELIVERY_ATTEMPT, deliveryAttempt);
}
```

So, `enhanceHeaders` must be fixed instead. Isn't it ?

Reply:
Ah; so that's the regression; we always need the the attributes now.

Reply:
https://github.com/spring-projects/spring-integration/issues/2958

Reply:
This is the regression:

https://github.com/spring-projects/spring-integration-kafka/commit/a50850ec9812181586956ea47092591e3ed1fd5f